include/swift/SIL/SILInstruction.h

```diff
@@ -87,6 +88,121 @@ SILInstructionKind getSILInstructionKind(StringRef InstName);
 /// Map SILInstructionKind to a corresponding SILInstruction name.
 StringRef getSILInstructionName(SILInstructionKind Kind);
 
+/// A formal SIL reference to a list of values, suitable for use as the result
+/// of a SILInstruction.
+///
+/// *NOTE* Most multiple value instructions will not have many results, so if we
+/// want we can cache up to 3 bytes in the lower bits of the value.
+///
+/// *NOTE* Most of this defined out of line further down in the file to work
+/// around forward declaration issues.
+class SILInstructionResultArray {
+  const SILInstruction *ParentInst;
```

Comment:
Hmm.  This is an interesting approach, but I think it would be better if the dynamic-dispatch costs were paid once, to get the array in the first place, rather than on every access to the array.  I like the idea of just storing an index in the iterator, but maybe we can store some sort of ValueBase* and length in SILInstructionResultArray?

Reply:
Ok. I will do an experiment.